ethereum/statetransition/src/main/java/tech/pegasys/artemis/statetransition/LmdGhost.java

```diff
@@ -0,0 +1,155 @@
+/*
+ * Copyright 2019 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.artemis.statetransition;
+
+import static java.util.Objects.requireNonNull;
+
+import com.google.common.primitives.UnsignedLong;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import net.consensys.cava.bytes.Bytes;
+import tech.pegasys.artemis.datastructures.blocks.BeaconBlock;
+import tech.pegasys.artemis.datastructures.operations.Attestation;
+import tech.pegasys.artemis.datastructures.state.BeaconState;
+import tech.pegasys.artemis.datastructures.util.BeaconStateUtil;
+import tech.pegasys.artemis.datastructures.util.ValidatorsUtil;
+import tech.pegasys.artemis.storage.ChainStorageClient;
+
+public class LmdGhost {
+
+  public BeaconBlock lmd_ghost(
+      ChainStorageClient store, BeaconState start_state, BeaconBlock start_block)
+      throws StateTransitionException {
+    List<Integer> active_validator_indices =
+        ValidatorsUtil.get_active_validator_indices(
+            start_state.getValidator_registry(),
+            BeaconStateUtil.slot_to_epoch(UnsignedLong.valueOf(start_block.getSlot())));
+
+    List<BeaconBlock> attestation_targets = new ArrayList<>();
+    for (Integer validatorIndex : active_validator_indices) {
+      try {
+        attestation_targets.add(get_latest_attestation_target(store, validatorIndex));
+      } catch (Error err) {
+        System.err.println(err);
+      }
+    }
+
+    BeaconBlock head = start_block;
+    List<BeaconBlock> children;
+    while (true) {
+      children = get_children(store, head);
+
+      if (children.size() == 0) {
+        return head;
+      }
+
+      head =
+          children.stream()
+              .max(
+                  Comparator.comparing(
+                      child_block ->
+                          Math.toIntExact(
+                              get_vote_count(store, child_block, attestation_targets).longValue())))
+              .get();
```

Comment:
that's gangster

Reply:
haha yep stream API I'm coming for you!!!